cmd/shell.go

```diff
@@ -276,6 +289,57 @@ func newStatsRegistry(addr string, logger blog.Logger) prometheus.Registerer {
 	return registry
 }
 
+// newOpenTelemetry sets up our OpenTelemetry tracing
+// It returns a graceful shutdown function to be deferred.
+func newOpenTelemetry(serviceName string, config OpenTelemetryConfig) func() {
+	r, err := resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String(serviceName),
+			semconv.ServiceVersionKey.String(core.GetBuildID()),
+		),
+	)
+	if err != nil {
+		FailOnError(err, "Could not create OpenTelemetry resource")
+	}
+
+	// Use a ParentBased sampler to respect the sample decisions on incoming
+	// traces, and TraceIDRatioBased to randomly sample new traces.
+	sampler := trace.TraceIDRatioBased(config.SampleRatio)
+	if !config.DisableParentSampler {
+		sampler = trace.ParentBased(sampler)
+	}
+
+	opts := []trace.TracerProviderOption{
+		trace.WithResource(r),
+		trace.WithSampler(sampler),
+	}
+
+	if config.Endpoint != "" {
+		exporter, err := otlptracegrpc.New(
+			context.Background(),
+			otlptracegrpc.WithInsecure(),
+			otlptracegrpc.WithEndpoint(config.Endpoint))
+		if err != nil {
+			FailOnError(err, "Could not create OpenTelemetry OTLP exporter")
+		}
+
+		opts = append(opts, trace.WithBatcher(exporter))
+	}
+
+	tracerProvider := trace.NewTracerProvider(opts...)
+	otel.SetTracerProvider(tracerProvider)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
+
+	return func() {
+		err := tracerProvider.Shutdown(context.Background())
```

Comment:
Some places, such as [here](https://github.com/letsencrypt/boulder/blob/e55a276efef324adf20c6bb47a539b8d1f2473b8/cmd/boulder-wfe2/main.go#L560-L561), we make sure the shutdown operation has a timeout. Probably a good idea here too.

Reply:
nice.  The tracerProvider.Shutdown() function here already takes a context for that purpose, so I'll plumb it up and add the shutdown timeout context.